forc-util/src/lib.rs

```diff
@@ -716,43 +716,104 @@ fn construct_window<'a>(
     let total_lines_of_highlight = end.line - start.line;
     debug_assert!(total_lines_in_input >= total_lines_of_highlight);
 
-    let mut current_line = 0;
-    let mut lines_to_start_of_snippet = 0;
-    let mut calculated_start_ix = None;
-    let mut calculated_end_ix = None;
-    let mut pos = 0;
-    for character in input.chars() {
+    let mut current_line = 1usize;
+
+    let mut chars = input.char_indices().map(|(char_offset, character)| {
+        let r = (current_line, char_offset);
         if character == '\n' {
-            current_line += 1
+            current_line += 1;
         }
+        r
+    });
 
-        if current_line + NUM_LINES_BUFFER >= start.line && calculated_start_ix.is_none() {
-            calculated_start_ix = Some(pos);
-            lines_to_start_of_snippet = current_line;
-        }
+    // Find the first char of the first line
+    let first_char = chars
+        .by_ref()
+        .find(|(current_line, _)| current_line + NUM_LINES_BUFFER >= start.line);
 
-        if current_line >= end.line + NUM_LINES_BUFFER && calculated_end_ix.is_none() {
-            calculated_end_ix = Some(pos);
-        }
+    // Find the last char of the last line
+    let last_char = chars
+        .by_ref()
+        .find(|(current_line, _)| *current_line > end.line + NUM_LINES_BUFFER)
+        .map(|x| x.1);
 
-        if calculated_start_ix.is_some() && calculated_end_ix.is_some() {
-            break;
+    // this releases the borrow of `current_line`
+    drop(chars);
+
+    let (first_char_line, first_char_offset, last_char_offset) = match (first_char, last_char) {
+        // has first and last
+        (Some((first_char_line, first_char_offset)), Some(last_char_offset)) => {
+            (first_char_line, first_char_offset, last_char_offset)
         }
-        pos += character.len_utf8();
+        // has first and no last
+        (Some((first_char_line, first_char_offset)), None) => {
+            (first_char_line, first_char_offset, input.len())
+        }
+        // others
+        _ => (current_line, input.len(), input.len()),
+    };
+
```

Comment:
I'm not sure about others, and although the code looks cleaner, I'm not sure we want to iterate over the whole string multiple times vs the old code that loops once but does multiple things during each iteration.

Reply:
The current code iterate only once for everything.
The older code apart from iterating once to find the window, was also iterating two more times inside the "window", given that `chars` in `O(n)`.

```
*start_ix = input[calculated_start_ix..(calculated_start_ix + start_ix_bytes)]
        .chars()
        .count();
    *end_ix = input[calculated_start_ix..(calculated_start_ix + end_ix_bytes)]
        .chars()
        .count();
```